examples/weakref.cpp

```diff
@@ -0,0 +1,279 @@
+  }
+
+ private:
+  using JobQueueStorage = JS::GCVector<JSObject*, 0, js::SystemAllocPolicy>;
+  JS::PersistentRooted<JobQueueStorage> m_queue;
+
+  using FunctionVector = JS::GCVector<JSFunction*, 0, js::SystemAllocPolicy>;
+  JS::PersistentRooted<FunctionVector> m_finalizationRegistryCallbacks;
+
+  // True if we are in the midst of draining jobs from this queue. We use this
+  // to avoid re-entry (nested calls simply return immediately).
+  bool m_draining : 1;
+
+  class SavedQueue : public JobQueue::SavedJobQueue {
+   public:
+    SavedQueue(JSContext* cx, CustomJobQueue* jobQueue)
+        : m_jobQueue(jobQueue),
+          m_saved(cx, std::move(jobQueue->m_queue.get())),
+          m_draining(jobQueue->m_draining) {}
+
+    ~SavedQueue() {
+      m_jobQueue->m_queue = std::move(m_saved.get());
+      m_jobQueue->m_draining = m_draining;
+    }
+
+   private:
+    CustomJobQueue* m_jobQueue;
+    JS::PersistentRooted<JobQueueStorage> m_saved;
+    bool m_draining : 1;
+  };
+
+  // JS::JobQueue override
+  js::UniquePtr<JS::JobQueue::SavedJobQueue> saveJobQueue(
+      JSContext* cx) override {
+    auto saved = js::MakeUnique<SavedQueue>(cx, this);
+    if (!saved) {
+      // When MakeUnique's allocation fails, the SavedQueue constructor is never
+      // called, so this->queue is still initialized. (The move doesn't occur
+      // until the constructor gets called.)
+      JS_ReportOutOfMemory(cx);
+      return nullptr;
+    }
+
+    m_queue.clear();
+    m_draining = false;
+    return saved;
+  }
+
+  bool maybeRunFinalizationRegistryCallbacks(JSContext* cx) {
+    bool ranCallbacks = false;
+
+    JS::Rooted<FunctionVector> callbacks{cx};
+    std::swap(callbacks.get(), m_finalizationRegistryCallbacks.get());
+    for (JSFunction* f : callbacks) {
+      JS::ExposeObjectToActiveJS(JS_GetFunctionObject(f));
+
+      JSAutoRealm ar{cx, JS_GetFunctionObject(f)};
+      JS::Rooted<JSFunction*> func{cx, f};
+      JS::Rooted<JS::Value> unused_rval{cx};
+      if (!JS_CallFunction(cx, nullptr, func, JS::HandleValueArray::empty(),
+                           &unused_rval)) {
+        LogPendingException(cx);
+      }
+
+      ranCallbacks = true;
+    }
+
+    return ranCallbacks;
+  }
+};
+
+static void CleanupFinalizationRegistry(JSFunction* callback,
+                                        JSObject* incumbent_global
+                                        [[maybe_unused]],
+                                        void* user_data) {
+  // Queue a cleanup task to run after each microtask has been run.
+  // We only have one global so ignore the incumbent global parameter.
+  auto* jobQueue = static_cast<CustomJobQueue*>(user_data);
+  jobQueue->queueFinalizationRegistryCallback(callback);
+}
+
+static bool GC(JSContext* cx, unsigned argc, JS::Value* vp) {
+  JS::CallArgs args = JS::CallArgsFromVp(argc, vp);
+
+  // FIXME: Should ClearKeptObjects be called anywhere in runJobs()?
+  JS::ClearKeptObjects(cx);
```

Comment:
I'm not sure if this is in the right place. Should this be called as part of the job queue also?

Reply:
This function does 3 separate things, where they're supposed to happen in different timing in practice:
  * (1) perform [ClearKeptObjects](https://tc39.es/ecma262/#sec-clear-kept-objects) by `JS::ClearKeptObjects` to remove `obj` from `[[KeptAlive]]` field, so that `obj` will be collected in the next GC call
    * In practice, this is supposed to happen after executing job ([microtask checkpoint also does so after draining job queue](https://whatpr.org/html/4571/webappapis.html#perform-a-microtask-checkpoint)), and [js::RunJobs does so](https://searchfox.org/mozilla-central/rev/9bb5d5f55da6cc7163dd93825c25609b769822f9/js/src/vm/JSContext.cpp#801,804). Thus, draining job queue by `js::RunJobs` is another (and more closer to the actual case) option
  * (2) perform GC by `JS_GC`, which collects `obj` and:
    *  set the `WeakRef`'s `[[WeakRefTarget]]` to EMPTY (https://tc39.es/ecma262/#sec-weakref-execution step 1.a.1)
       * at this point, `weakRef.deref()` becomes `undefined`
    *  enqueues `FinalizationRegistry` cleanup job for `obj` to the job queue (https://tc39.es/ecma262/#sec-weakref-execution step 1.b.ii)
  * (3) drain job queue  by `js::RunJobs`, and run the `FinalizationRegistry` cleanup job
    * at this point `valueFinalized` is set to `"marker"`

So, the "GC" name sounds misleading, and I think it's better having 3 separate functions (or 2 if (1) is done by `js::RunJobs`) for each and make it clearer why each of them are necessary in the example.

If step (2) and (3) above becomes separate function, `if (weakRef.deref() !== undefined) throw new Error("WeakRef");` can be put between them, so that it also illustrates what happens at which point.

Reply:
This is great advice. I understand it a lot better after your explanation. I made 2 separate functions (one that only calls `js::RunJobs` and one that only calls `JS_GC`) and added comments inside the snippet of JS code.